WIP: Returning non_deterministic column must match actually persisted value

BaurzhanSakhariev opened the pull request:
Closes https://github.com/crate/crate/issues/14748

BaurzhanSakhariev commented on server/src/main/java/io/crate/execution/dml/Indexer.java:
```diff
@@ -106,8 +99,11 @@ public class Indexer {
-    private final List<Input<?>> returnValueInputs;
+    private final Symbol[] returnValues;
+    private final Context<CollectExpression<IndexItem, Object>> ctxForRefs;
     private final List<Synthetic> undeterministic = new ArrayList<>();
+    private final Set<ColumnIdent> undeterministicIdents = new HashSet<>();
```
we can use smth like `Symbols.isDeterministic(ref.gen / ref.def())` but we need to on indexing each item, creating small Set should be fine - normally a table wouldn't contain many non-deterministic columns.

BaurzhanSakhariev commented on server/src/main/java/io/crate/execution/dml/Indexer.java:
```diff
@@ -488,14 +486,7 @@ public Indexer(String indexName,
-                this.returnValueInputs.add(ctxForRefs.add(returnValue));
-            }
-        }
+        this.returnValues = returnValues;
```
This is now deferred - hence `returnValues` and `ctxForRefs` are fields of the Indexers, We need to replace refs and only then create new inputs.

BaurzhanSakhariev commented on server/src/main/java/io/crate/execution/dml/Indexer.java:
```diff
@@ -661,15 +665,40 @@ public ParsedDocument index(IndexItem item) throws IOException {
+
+        for (int i = 0; i < returnValues.length; i++) {
+            Symbol updatedReturnValue = RefReplacer.replaceRefs(returnValues[i], ref -> {
+                if (undeterministicIdents.contains(ref.column())) {
+                    Object value = undeterministicValues.get(ref.column());
+                    return Literal.ofUnchecked(ref.valueType(), value);
+                }
```
We are just using stored computed values now.

In `legacyInsert` (which is removed now) we [passed source](https://github.com/crate/crate/commit/672a882a82a442cdb007fdfe7eac9391d17cf5e6#diff-afc8b63b41c53d6e6ea5b956bdd39557ab55e1dfe1b3e1242f1d4e29a9fb57a0L607) (Doc) and parsed it via [DocRefResolver expressions](https://github.com/crate/crate/commit/672a882a82a442cdb007fdfe7eac9391d17cf5e6#diff-ec600b9f2d9172b3dd21c8acc88623c601fa3cc4db2ed3986682b35308024089L45) so this is still lightweight.

BaurzhanSakhariev commented on server/src/test/java/io/crate/execution/dml/IndexerTest.java:
```diff
@@ -561,6 +560,39 @@ public void test_can_generate_return_values() throws Exception {
         assertThat(returnValues).containsExactly("dummy-id-1", 10, 20);
     }
 
+    @Test
+    public void test_return_values_referring_to_non_deterministic_columns_use_persisted_value() throws Exception {
+        SQLExecutor e = SQLExecutor.builder(clusterService)
+            .addTable("create table tbl (x int, y int default round((random() + 1) * 100), z int as round((random() + 1) * 100))")
+            .build();
+
+        DocTableInfo table = e.resolveTableInfo("tbl");
```
I confirmed that reverting fix make this test fail. 

Should I remove integration test?

matriv replied:
Although adding itests is not preferrable, If we don't have another one to test this behaviour I'd prefer to have it as well on top of the unit test of Indexer.

matriv commented on server/src/main/java/io/crate/execution/dml/Indexer.java:
```diff
@@ -585,18 +585,23 @@ public ParsedDocument index(IndexItem item) throws IOException {
-                ValueIndexer<Object> indexer = synthetic.indexer();
+
                 Object value = synthetic.input().value();
+                if (undeterministicIdents.contains(column)) {
```
`ObjectIndexer#indexValue()#L115` is also calling: `innerValue = synthetic.input().value();` shouldn't we store and use this as well?
Can you please add test with returning a subcol non-deterministic value?

BaurzhanSakhariev replied:
Good point, let me check it

BaurzhanSakhariev replied:
That's a bug indeed, thanks for catching!

I pushed a test update (both unit/itest) will provide a fix.

matriv replied:
Could we do the other way around?
Can we store already all generated/default values in the ctor, and use them for indexing when required? Maybe we need to pass a different structure through `indexValue` but I'm not sure how this will work if actual index values are used in those expressions to calculate the final generated value. (Keep in mind the issue is the same both for `default` and `generated as....`.

BaurzhanSakhariev replied:
If we store gen/default columns in ctor we will get same values for such columns across the whole table which will be a bug.

We even had this issue and it got fixed in https://github.com/crate/crate/pull/12126/files#diff-6b4ff31a02425d333cc54423875ca7ab94432d3399bd0d6b45797763cbb5dcd7R86-R89

matriv replied:
I just don't like that in `getImplementation()` we already calculate those values:
https://github.com/crate/crate/blob/master/server/src/main/java/io/crate/execution/dml/Indexer.java#L190
https://github.com/crate/crate/blob/master/server/src/main/java/io/crate/execution/dml/Indexer.java#L196
https://github.com/crate/crate/blob/master/server/src/main/java/io/crate/execution/dml/Indexer.java#L222
and then during indexing we need to calculate again.

matriv replied:
> If we store gen/default columns in ctor we will get same values for such columns across the whole table which will be an issue.
> 
> We even had this issue and it got fixed in https://github.com/crate/crate/pull/12126/files#diff-6b4ff31a02425d333cc54423875ca7ab94432d3399bd0d6b45797763cbb5dcd7R86-R89
> 
> We have to compute synthetics on each `index()` but now we are also caching/reusing non-deterministic values.

Makes sense, thx!

BaurzhanSakhariev replied:
I pushed a fixup(s), could you take another look? `@Repeat` is temporal, I will revert it (or maybe let's drop Itest?)

BaurzhanSakhariev replied:
sorry, just saw another comment to keep itest, will only remove `repeat` then

mfussenegger reviewed (commented):
Couldn't we cache the value in the synthetic object and invalidate/override on a new index call? 

I don't think it's necessary to add new parameters to the valueIndexer and it would be possible to keep this more encapsulated

BaurzhanSakhariev commented:
Converting to the draft - integration test is flaky (I suspect indexing on replica logic needs to be double checked)

BaurzhanSakhariev commented:
> Couldn't we cache the value in the synthetic object and invalidate/override on a new index call?

Thanks, tried it out and it looks simpler. Some notes:

Since we anyway have a field for computed value,  all synthetics are stored (deterministic ones as well). It's easier because:

1. We don't recompute even deterministic functions on `returnValues()`
2. We don't have to differentiate between deterministic or not .ie don't have to introduce new flag in `Synthetics` or again pass through something like `nonDetermisiticIdents` through all `ValueIndexer`-s
3. We don't have to clear anything between `index()` calls since we always explicitly recompute.

Looks like there was another bug on master which is fixed now as well:
New test was flaky and was fixed by using cached values for replica in https://github.com/crate/crate/pull/14762/commits/b32f7265ad8745018c754167904823fed9ffe2d6.

On master we call `synthetic.input.value()` twice but test `test_generated_non_deterministic_value_is_consistent_on_primary_and_replica` passes. 

This because now() is "half" non-determinsitc, it returns same values within same query scope. I updated test to use ransom which fails on master and passes here. Added a change entry as well.

mfussenegger commented on server/src/main/java/io/crate/execution/dml/Indexer.java:
```diff
@@ -661,15 +693,46 @@ public ParsedDocument index(IndexItem item) throws IOException {
+        // We must not change original RETURNING symbols.
+        // Otherwise, RefReplacer.replaceRefs will changes Refs to Literals on the first index()
+        // and all subsequent calls will not find a Reference and will operate with old Literals.
+        Symbol[] updatedReturnValues = new Symbol[returnValues.length];
+
+        for (int i = 0; i < returnValues.length; i++) {
+            Symbol updatedReturnValue = RefReplacer.replaceRefs(returnValues[i], ref -> {
+                Synthetic synthetic = synthetics.get(ref.column());
+                if (synthetic != null) {
+                    // Deterministic values are also reused just in order not to compute twice.
+                    return Literal.ofUnchecked(ref.valueType(), synthetic.computedValue());
+                }
+                return ref;
+            });
+            updatedReturnValues[i] = updatedReturnValue;
+        }
+
+        List<Input<?>> returnValueInputs = new ArrayList<>(returnValues.length);
+        for (Symbol returnValue : updatedReturnValues) {
+            returnValueInputs.add(ctxForRefs.add(returnValue));
+        }
```
I suspect re-creating inputs for the return values on each row may have quite a bit of impact on performance.

Couldn't we let `Synthetic` implement `Input` and use them within the `returnValueInputs`, so we don't need to re-create them per row?

That should also allow to fully encapsulate the cache on a `value()` call. Right now consumers have to know when to call `value()` and when to call `computedValue()` or activate the cache.

Downside is that we'd have to bring back invalidation.

BaurzhanSakhariev replied:
Thanks for the suggestion!

Implemented in https://github.com/crate/crate/pull/14762/commits/529b01e17ca57bdbc1a80bc601883ed6dec0b3ea

mfussenegger commented on server/src/main/java/io/crate/execution/dml/Indexer.java:
```diff
@@ -661,15 +693,46 @@ public ParsedDocument index(IndexItem item) throws IOException {
         }
     }
 
+    /**
+     *
+     * RETURNING can refer to multiple non-deterministic columns,
+     * generated and defaults and their combinations, e.g., gen_col + def_col.
+     *
+     * We need to ensure that we use already computed values for such references,
+     * this this must be called after {@link #index(IndexItem)} have been called.
+     *
+     */
```
```suggestion
```

Method docs should be primarily for API consumers. This is addressing the implementation. I think we should move this inside the method body but see also my comment below. This might even become unnecessary if we change the Synthetic to an Input

mfussenegger commented on server/src/main/java/io/crate/execution/dml/Indexer.java:
```diff
@@ -523,7 +533,24 @@ public Indexer(String indexName,
+                Symbol updatedReturnValue = RefReplacer.replaceRefs(returnValues[i], ref -> {
+                    Synthetic synthetic = synthetics.get(ref.column());
+                    if (synthetic != null) {
+                        return Literal.ofUnchecked(ref.valueType(), synthetic.value());
```
Doesn't this freeze the value at construction time? 
We'd need to get the `returnValueInputs` to call into the synthethic. That means either we've to override the `Literal#value` call here, or we'd have to change the `InputFactory` stuff to return different inputs for the functions.

BaurzhanSakhariev replied:
Yep, I have something different locally, created anther factory to either get synthetic or getImplementation

BaurzhanSakhariev replied:
Replied in https://github.com/crate/crate/pull/14762#discussion_r1342532437

mfussenegger commented on server/src/main/java/io/crate/execution/dml/Indexer.java:
```diff
@@ -239,9 +239,50 @@ private Object fromGenerated(GeneratedReference generated, IndexItem item) {
+
+        @Override
+        public Object value() {
+            if (computedValue == null) {
```
I think this needs a dedicated flag to check if it is computed, in case a non-deterministic function can return null or a value.

BaurzhanSakhariev replied:
Added flag in https://github.com/crate/crate/pull/14762/commits/fa99342bc6cae62ee9fcff481a63878be5a453fb.

Looks like we currently don't have non-deterministic functions, returning value or null - do you mean we need such flag to be future proof?

mfussenegger replied:
>  do you mean we need such flag to be future proof in case we add such a function?

Or user defined functions

mfussenegger commented on server/src/main/java/io/crate/execution/dml/Indexer.java:
```diff
@@ -491,9 +532,23 @@ public Indexer(String indexName,
+            Context<Input<?>> ctxForReturnValues = inputFactory.ctxForRefs(
+                txnCtx,
+                ref -> {
+                    Synthetic synthetic = synthetics.get(ref.column());
+                    if (synthetic == null) {
+                        // We must not use ReferenceResolver.getImplementation(ref) here, since Input won't be added to the expressions.
+                        // Instead, we add Reference to the existing context, to make sure that Input is added to the expressions.
+                        // Also, reference won't be added twice to the context (if referred twice), since InputFactory caches inputs per reference.
+                        return ctxForRefs.add(ref);
+                    } else {
+                        return synthetic;
+                    }
+                }
```
```suggestion
                    // Using Synthethic if available, it caches results to ensure non-deterministic functions yield the same result
                    // across indexing and return values
                    Synthetic synthetic = synthetics.get(ref.column());
                    if (synthetic == null) {
                        return ctxForRefs.add(ref);
                    } else {
                        return synthetic;
                    }
                }
```
